### ISSUE TITLE
CA1508 false positive with for-loop on .NET 5

### ISSUE THREAD

**Issue by AtOMiCNebula** [dotnet/roslyn-analyzers#4637]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

.NET 5.0.101 SDK (and VS 2019 v16.8.3), repros with or without NetAnalyzers v5.0.1 package manually installed.

### Describe the bug

I was writing out some code for a personal project using the new .NET 5 SDKs that were just released, and have been enabling a number of the code analyzers.  While writing some code, I ran into an instance of CA1508 that is clearly incorrect.

```csharp
public static void Main()
{
    HashSet<int> state = new HashSet<int>
    {
        0,
    };

    int hits = 0;
    for (int i = 0; i < 1; i++)
    {
        hits += state.Contains(i) ? 1 : 0;
    }

    if (hits != 0) // CA1508 on this line, says 'hits != 0' will always be false
    {
        Console.WriteLine($"Analyzer thinks we will never have any hits, but we had {hits}...");
    }
}
```

Handy repro repo: https://github.com/AtOMiCNebula/CA1508FalseAlarm/

### Steps To Reproduce

Compile the above code in a .NET 5.0 project with `<AnalysisMode>AllEnabledByDefault</AnalysisMode>` set in the .csproj.

### Expected behavior

No warning on the `hits != 0` condition.

### Actual behavior

Analyzer complains that `hits != 0` will always be false, even though running the code shows the opposite.

## Additional context

I've managed to reduce the code pretty far from where I started out with it, but the `for` loop seems to be important here.  Hopefully this is reduced far enough to be helpful. :smile:

**Comment by mavasani** [comment]
Verified that this issue is fixed with one of the latest PRs. It repros on latest .NET8 preview SDK, but not on the latest NuGet package: https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet8/NuGet/Microsoft.CodeAnalysis.NetAnalyzers/overview/8.0.0-preview.23402.3.

Fix should be part of the upcoming .NET8 preview SDK
